code/scripting/scripting.cpp

```diff
@@ -1137,10 +1141,34 @@ bool script_state::ParseCondition(const char *filename)
 	return true;
 }
 
-void script_state::AddConditionedHook(ConditionedHook hook) { ConditionalHooks.push_back(std::move(hook)); }
+void script_state::AddConditionedHook(ConditionedHook hook) {
+	ConditionalHooks.push_back(std::move(hook));
+	AssayActions();
+}
 
 void script_state::AddGameInitFunction(script_function func) { GameInitFunctions.push_back(std::move(func)); }
 
+// For each possible script_action this maintains an array that records whether any scripts are actually using this action
+// This allows us to avoid significant overhead from checking everything at the potential hook sites, but you must call
+// AssayActions() after modifying ConditionalHooks before returning to normal operation of the scripting system!
+void script_state::AssayActions() {
+	for (bool &i : ActiveActions) {
+		i = false;
+	}
+
+	for (const auto &hook : ConditionalHooks) {
+		for (const auto &action : hook.Actions) {
+			if (action.action_type >= -1 && action.action_type <= CHA_LAST) {
+				ActiveActions[action.action_type] = true;
+			}
+		}
+	}
+}
```

Comment:
Why does this function clear the entire array and then re-populate it whenever it's called?  Why not initialize the array when the script system is initialized, and then just toggle an action to true when a hook is added?

Reply:
seems like this would be fine if this only gets called once, after all the script tables have been loaded? I'm failing right now to imagine  a case where this needs to be done more than once.

Reply:
It also had to be run in AddConditionedHook:

https://github.com/scp-fs2open/fs2open.github.com/pull/3538/files#diff-8763caafc296d3dd73ccf6d72b40529289c3908b57fa9f06da1fc26ac840f547R1146

I also added it to the Clear method on script_system, but that was probably overkill as I think that's one of those dumb cleanup functions we have that only actually gets called when the game is about to exit.

As for why I did it this way, I just found it easier to write a function that would guarantee the active action lookup was accurate and then drop it in between any code that might add new conditional hooks and its return to the scripting system, versus fiddling around with various individual sites that might add hooks. It may look a little inefficient but consider that script_state::RunCondition iterates over every single conditional hook and does even more stuff than this hundreds or thousands of times per frame. AssayActions is probably run a few dozen times on engine startup at most.